### PR TITLE
Add loading script to load experiment design data into database

### PR DESCRIPTION
**Pull request by haideriqbal**
We want to get away with the experiment design file and in order to do so, we created a schema consisting of two tables in the database. Details of the schema can be found in this [PR](https://github.com/ebi-gene-expression-group/atlas-schemas/pull/26).

We are loading data from two files one `condensed_sdrf` and a column sequence from `sdrf` file. Once the data is correctly loaded in the database then we will move on to the web application part to stream the data from the db to show on the website.

### CODE REVIEW

File: bin/load_exp_design.sh [hunk 1]
```diff
@@ -0,0 +1,30 @@
...
+source $scriptDir/db_scxa_common.sh
+
+dbConnection=${dbConnection:-$1}
+condensed_sdrf_file=${CONDENSED_SDRF_FILE:-$2}
+sdrf_file=${SDRF_FILE:-$3}
+
+# Check that necessary environment variables are defined.
+[ -z ${dbConnection+x} ] && echo "Env var dbConnection for the database connection needs to be defined. This includes the database name." && exit 1
+[ -z ${CONDENSED_SDRF_FILE+x} ] && echo "Env var CONDENSED_SDRF_FILE for the experiment design data needs to be defined." && exit 1
+[ -z ${SDRF_FILE+x} ] && echo "Env var SDRF_FILE for column sequence of experiment design needs to be defined." && exit 1
```
Comment (pmb59) [inline bin/load_exp_design.sh:15]: If these checks are OK, then they should be above L8.

Or are you intending to check here `condensed_sdrf_file` and `sdrf_file`?

**Review by pmb59 — CHANGES_REQUESTED** [review]
Hi @haideriqbal Would it possible to add a test for this new function? 
I think you can export variables in `tests/run_tests.sh` and add a test here `tests/random-data-set.bats`

File: bin/load_exp_design.sh (hunk 1 above)
Comment (haideriqbal) [reply bin/load_exp_design.sh:15]: I followed the convention data production team has been using in the scripts. For example, [here](https://github.com/ebi-gene-expression-group/db-scxa/blob/develop/bin/add_exps_to_collection.sh) and [here](https://github.com/ebi-gene-expression-group/db-scxa/blob/develop/bin/load_db_scxa_analytics.sh).

I do not have a strong opinion about this was just following the format found in other scripts.

**Review by pmb59 — CHANGES_REQUESTED** [review]
Hi @haideriqbal 
I have still one comment: In [3f7b1bd](https://github.com/ebi-gene-expression-group/db-scxa/pull/63/commits/3f7b1bd7d469993eb80e036af8abd66f8dd9b9c0) you removed this chunk

```
done < $condensed_sdrf_file
```
so the question now is where is `$condensed_sdrf_file` used? (and what is the `while` reading)

**Comment by haideriqbal** [comment]
> Hi @haideriqbal I have still one comment: In [3f7b1bd](https://github.com/ebi-gene-expression-group/db-scxa/pull/63/commits/3f7b1bd7d469993eb80e036af8abd66f8dd9b9c0) you removed this chunk
> 
> ```
> done < $condensed_sdrf_file
> ```
> 
> so the question now is where is `$condensed_sdrf_file` used? (and what is the `while` reading)

Good spot! It was deleted by mistake 🤕 reverted it now!

File: bin/load_exp_design.sh [hunk 2]
```diff
@@ -0,0 +1,38 @@
...
+        else
+            for element in "${ARRAY[@]}"; do
+              search_column="$element[${col_name}]"
+              column_order=$(awk -v val="$search_column" -F '\t' '{for (i=1; i<=NF; i++) if ($i==val) {print i} }' $sdrf_file)
+            done
+        fi
+        echo "INSERT INTO exp_design_column (experiment_accession, column_name, sample_type, column_order) VALUES ('$exp_acc', '$col_name', '$sample_type', '$column_order');" | psql -v ON_ERROR_STOP=1 $dbConnection
+        echo "INSERT INTO exp_design (sample, annot_value, annot_ont_uri, exp_design_column_id) VALUES ('$sample', '$annot_value', '$annot_url', (SELECT id FROM exp_design_column WHERE experiment_accession='$exp_acc' AND column_name='$col_name' AND sample_type='$sample_type'));" | psql -v ON_ERROR_STOP=1 $dbConnection
+
+done
```
Comment (pmb59) [inline bin/load_exp_design.sh:39]: it looks like we are missing again the `done < $condensed_sdrf_file`?

File: bin/load_exp_design.sh [hunk 3]
```diff
@@ -0,0 +1,38 @@
...
+sdrf_file=${SDRF_FILE:-$3}
+
+# Check that necessary environment variables are defined.
+[ -z ${dbConnection+x} ] && echo "Env var dbConnection for the database connection needs to be defined. This includes the database name." && exit 1
+[ -z ${CONDENSED_SDRF_FILE+x} ] && echo "Env var CONDENSED_SDRF_FILE for the experiment design data needs to be defined." && exit 1
+[ -z ${SDRF_FILE+x} ] && echo "Env var SDRF_FILE for column sequence of experiment design needs to be defined." && exit 1
+
+# Reason for creating this array is to search factor value column
+# In some sdrf files this column is mentioned as "Factor Value" and in some as "FactorValue"
+ARRAY=( FactorValue "Factor Value" )
```
Comment (pmb59) [inline bin/load_exp_design.sh]: I would rename `ARRAY` to FactorArray, FactorValueArray or similar

File: bin/load_exp_design.sh [hunk 4]
```diff
@@ -0,0 +1,38 @@
...
+do
+        if [ $sample_type == 'characteristic' ]
+        then
+            search_column="Characteristics[${col_name}]"
+            column_order=$(awk -v val="$search_column" -F '\t' '{for (i=1; i<=NF; i++) if ($i==val) {print i} }' $sdrf_file)
+        else
+            for element in "${ARRAY[@]}"; do
+              search_column="$element[${col_name}]"
+              column_order=$(awk -v val="$search_column" -F '\t' '{for (i=1; i<=NF; i++) if ($i==val) {print i} }' $sdrf_file)
+            done
```
Comment (pmb59) [inline bin/load_exp_design.sh]: ARRAY will take the values `FactorValue` and `"Factor Value" `.

If in a SDRF file this column is mentioned as `FactorValue`, then the second loop in the for (`"Factor Value" `)will produce empty values for `search_column` and `column_order`?

Maybe you should `break` the loop if valid value is found?

**Review by pmb59 — CHANGES_REQUESTED** [review]

File: bin/load_exp_design.sh (hunk 4 above)
Comment (haideriqbal) [reply bin/load_exp_design.sh]: great catch! added loop break

File: bin/load_exp_design.sh (hunk 3 above)
Comment (haideriqbal) [reply bin/load_exp_design.sh]: done!

File: bin/load_exp_design.sh (hunk 2 above)
Comment (haideriqbal) [reply bin/load_exp_design.sh:39]: surpisingly got missed 😮‍💨  added again

File: bin/load_exp_design.sh [hunk 5]
```diff
@@ -0,0 +1,46 @@
...
+        fi
+        echo "INSERT INTO exp_design_column (experiment_accession, column_name, sample_type, column_order) VALUES ('$exp_acc', '$col_name', '$sample_type', '$column_order');" | psql -v ON_ERROR_STOP=1 $dbConnection
+done
+
+while IFS=$'\t' read exp_acc sample sample_type col_name annot_value annot_url
+do
+  echo "INSERT INTO exp_design (sample, annot_value, annot_ont_uri, exp_design_column_id) VALUES ('$sample', '$annot_value', '$annot_url', (SELECT id FROM exp_design_column WHERE experiment_accession='$exp_acc' AND column_name='$col_name' AND sample_type='$sample_type'));" | psql -v ON_ERROR_STOP=1 $dbConnection
+done < $condensed_sdrf_file
+
+echo "Experiment design data done loading!"
```
Comment (pmb59) [inline bin/load_exp_design.sh]:
```suggestion
echo "Experiment design data done loading for $condensed_sdrf_file"
```
small suggestion to ease debugging

**Review by pmb59 — APPROVED** [review]
@haideriqbal LGTM!

**Pull request merged**